zkvm/src/types.rs

```diff
@@ -281,15 +317,25 @@ impl Data {
     pub fn len(&self) -> usize {
         match self {
             Data::Opaque(data) => data.len(),
+            // TBD: calculate length of witness data
+            // Is there a way to do this without just serializing the
+            // data and taking its length?
             Data::Witness(_) => unimplemented!(),
         }
     }
 
     /// Converts the Data into a vector of bytes
-    pub fn to_bytes(self) -> Vec<u8> {
+    pub fn to_bytes(&self) -> Vec<u8> {
         match self {
-            Data::Opaque(data) => data,
-            Data::Witness(_) => unimplemented!(),
+            Data::Opaque(data) => data.clone(),
+            Data::Witness(w) => {
+                let mut bytes: Vec<u8> = Vec::new();
+                // TBD: right now the encode/to_bytes flow
+                // seems a little backwards. Might be better
+                // to do it the other way around.
```

Comment:
yeah, lets have `encode(buf: &mut Vec<u8>)` as the primary interface for writing data, and in `to_bytes(self)` (still consuming self) use it to encode opaque data efficiently by moving Vec and for witness, simply using encode() the way it happens in this patch.